rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,159 @@
+
+#### Exposing Mapper-Like Postgres Interface
+
+We can put the mapper's logic inside Postgres. And then provide an interface for the engine to call mapper functions. In this approach, the interface will be provided through the Postgres, and therefore this is the approach that is most consistent with "Pure-SQL". The DSL idea from the above way can still be applied here (developer writes DSL, PL/pgSQL code is generated).
+
+So let's consider one possible implementation of the mapper structure (all as Postgres functions):
+- CloudId: `{entity}_cloud_id(object: json) -> string` a Postgres function that takes the object as json and returns its cloud id as string.
+- Equals: `{entity}_equal(o1: json, o2: json) -> boolean`
+- CRUD functions:
+    - `{entity}_cloud_create(object: json)`
+    - `{entity}_cloud_read() -> json list`
+    - `{entity}_cloud_update(old: json, new: json)`
+    - `{entity}_cloud_delete(object: json)`
+- Call other mappers: 
+  - `call_mapper('{module_name}', '{entity_name}', 'cloud|db', 'create|read|update|delete', {}::json) -> json`
+  - It would be easy to add support for normal modules with the above definition. It will call `ec2_cloud_read` if that Postgres function exists. Otherwise, it'll send an RPC to the engine to ask engine to call that function from the mapper.
+
+We can also consider the above functions to be strongly-typed inside Postgres. More info here:
+https://rounded-apology-58d.notion.site/What-Does-a-Mapper-Need-9402ce7e35244405876f4cf660365c86
+
+### Common Parts
+We also need these added functionalities in order for any of the above solutions to work:
+- Dynamically creating a mapper that provides the same interface as `MapperBase`, but calls corresponding SQL functions.
+- Registering/de-registering logic for the module entities be transmitted to the engine: security concerns here, since the module should not be able to `DROP` tables for other modules (for example).
+
+### Security Can't Be Guaranteed
+Suppose this scenario:
+> We have a new Pure-SQL module. It has access to do `INSERT`s and `SELECT`s on behalf of the user. So it can create a CodeBuild project that calls a curl, with the AWS credentials coming from the `aws_credentials` table, right? So I think security can't be guaranteed in neither of the above ways, given our tables keep the current permissioning system.
+
+One way to fix the above issue, is to create a new Postgres role for each user (like `{username}_module`) which will be used by the Pure-SQL modules to execute SQL commands. It won't have access to the `aws_credentials` table.
+
+But the above idea can be also bypassed:
+> Use `aws_sdk` module to create a new set of AWS AccessKeyId and SecretAccessKey. Then create a CodeBuild project to send them to the malicious endpoint :D.
```

Comment:
As we are discussing in chat, being explicit about module capabilities and dependencies up-front can help mitigate this. A third party module that depends on CodeBuild and AWS account key management APIs should declare that it needs these things and if they seem incongruous to the purpose of the module that's another flag for users.

We could also classify certain API calls as requiring manual confirmation because of their sensitivity (particularly the aws credential management APIs, but also anything that allows arbitrary code execution, such as codebuild or even the startup script you can give to an ec2 instance) as further hardening in the future.